File: src/main/java/me/syari/ss/discord/api/requests/RestFuture.java
package me.syari.ss.discord.api.requests;

import me.syari.ss.discord.internal.JDAImpl;
import me.syari.ss.discord.internal.requests.RestActionImpl;
import me.syari.ss.discord.internal.requests.Route;
import okhttp3.RequestBody;
import org.apache.commons.collections4.map.CaseInsensitiveMap;

import java.util.concurrent.CompletableFuture;

public class RestFuture<T> extends CompletableFuture<T> {
    final Request<T> request;

    public RestFuture(final RestActionImpl<T> restAction, final boolean shouldQueue,
                      final RequestBody data,
                      final Route.CompiledRoute route, final CaseInsensitiveMap<String, String> headers) {
        this.request = new Request<T>(restAction, this::complete, this::completeExceptionally, shouldQueue, data, route, headers);
        ((JDAImpl) restAction.getJDA()).getRequester().request(this.request);
    }

    @Override
    public boolean cancel(final boolean mayInterrupt) {
        if (this.request != null)
            this.request.cancel();

        return (!isDone() && !isCancelled()) && super.cancel(mayInterrupt);
    }
}

File: src/main/java/me/syari/ss/discord/api/requests/RestFuture.kt
package me.syari.ss.discord.api.requests

import me.syari.ss.discord.internal.requests.Requester
import me.syari.ss.discord.internal.requests.RestAction
import me.syari.ss.discord.internal.requests.Route
import okhttp3.RequestBody
import java.util.concurrent.CompletableFuture

class RestFuture<T>(
    restAction: RestAction<T>, shouldQueue: Boolean, data: RequestBody?, route: Route
): CompletableFuture<T>() {
    private val request: Request<T>?
    override fun cancel(mayInterrupt: Boolean): Boolean {
        request?.cancel()
        return !isDone && !isCancelled && super.cancel(mayInterrupt)
    }

    init {
        request = Request(
            restAction,
            { complete(it) },
            { completeExceptionally(it) },
            shouldQueue,
            data,
            route
        )
        Requester.request(request)
    }
}